File: day8/typescript/day8.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

interface Registers {
    [name: string]: number;
}

interface Instruction {
    register: string;
    amount: number;
    condition: string[];
}

const getRegister = (regs: Registers, reg: string) => R.compose<Registers, number, number>(
    R.defaultTo(0),
    R.prop(reg),
)(regs);

const parse = (ins: string) => {
    const [register, op, num, _, ...condition] = R.split(' ', ins);
    const amount = parseInt(num, 10);
    return {
        register,
        amount: op === 'inc' ? amount : -1 * amount,
        condition,
    };
};

const solve = (instructions: Instruction[]): [number, Registers] => {
    let max = 0;
    const regs: Registers = {};
    for (const { register, amount, condition } of instructions) {
        const [cond, op, val] = condition;
        if (eval(`${getRegister(regs, cond)} ${op} ${val}`)) {
            regs[register] = getRegister(regs, register) + amount;
            max = R.max(max, regs[register]);
        }
    }

    return [max, regs];
};

(async () => {
    const input = await promisify(readFile)('day8/input.txt', 'utf8');
    const [max, regs] = R.compose(
        solve,
        R.map(parse),
        R.split('\r\n'),
    )(input);

    console.log(Math.max(...R.values(regs)));
    console.log(max);
})();

File: day8/haskell/day8.hs
import Control.Arrow (first)
import Data.Map (Map, findWithDefault, empty, insert)

type State = (Map String Int, Int)
type Instruction = (String, Int, String, Int -> Bool)

op :: String -> (Int -> Int -> Bool)
op ">"  = (<)
op "<"  = (>)
op ">=" = (<=)
op "<=" = (>=)
op "==" = (==)
op "!=" = (/=)
op _    = error "Unknown operation"

parse :: String -> Instruction
parse s = (r, if m == "inc" then read n else -(read n), r', op o (read v))
    where [r, m, n, "if", r', o, v] = words s

modify :: State -> Instruction -> State
modify s@(rs, m) (r, a, r', c)
        | not . cond $ rs   = s
        | otherwise         = (insert r val rs, max m val)
    where cond = c . findWithDefault 0 r'
          val  = findWithDefault 0 r rs + a

solve :: [Instruction] -> (Int, Int)
solve = first maximum . foldl modify (empty, 0)

main :: IO ()
main = do
    is <- map parse . lines <$> readFile "day8/input.txt"
    print . solve $ is